### PR TITLE
[FREELDR] Use less memory when doing unaligned reads on Btrfs

### PR DESCRIPTION
**Pull request by maharmstone**
We're still in real mode, where memory is at a premium, when we call BtrFsRead, and I was getting out of memory errors. This changes it so that we only allocate a temp buffer for the first 512-byte sector, and read normally from then on.

### CODE REVIEW

**Comment by HBelusca** [comment]
"in real mode" ? Except when FreeLdr calls Int386() functions where it switches to real mode, it runs otherwise in 32-bit protected mode.

**Comment by maharmstone** [comment]
Thanks - I assumed it was because we only had 1MB of memory to play with! Which, now that I think about it, doesn't make much sense as the kernel is bigger than that...

I still needed this patch to get Freeldr working with XP, though - it fails otherwise when trying to allocate a few hundred kilobytes.

**Comment by HBelusca** [comment]
Note that, unless this has regressed pretty recently, that FreeLdr is capable of loading Server 2003, so I wouldn't expect a lot of differences wrt. loading XP.

File: boot/freeldr/freeldr/lib/fs/btrfs.c [hunk 1]
```diff
@@ -699,18 +699,32 @@ static u64 btrfs_read_extent_reg(struct btrfs_path *path, struct btrfs_file_exte
...
         if (physical & (512 - 1))
         {
+            u32 shift;
+
             /* If somebody tried to do unaligned access */
-            physical -= offset;
-            temp_out = FrLdrTempAlloc(size + offset, TAG_BTRFS_FILE);
 
-            if (!disk_read(physical, temp_out, size + offset))
+            temp_out = FrLdrTempAlloc(512, TAG_BTRFS_FILE);
```
Comment (SergeGautherie) [inline boot/freeldr/freeldr/lib/fs/btrfs.c]: Would `char temp_out[512]` on stack work?

File: boot/freeldr/freeldr/lib/fs/btrfs.c [hunk 2]
```diff
@@ -699,18 +699,32 @@ static u64 btrfs_read_extent_reg(struct btrfs_path *path, struct btrfs_file_exte
...
+
+            if (!disk_read(physical & ~(512 - 1), temp_out, 512))
```
Comment (SergeGautherie) [inline boot/freeldr/freeldr/lib/fs/btrfs.c]: Use something like `#define BTRFS_SECTOR_SIZE 512`?

File: boot/freeldr/freeldr/lib/fs/btrfs.c [hunk 3]
```diff
@@ -699,18 +699,32 @@ static u64 btrfs_read_extent_reg(struct btrfs_path *path, struct btrfs_file_exte
...
             {
                 FrLdrTempFree(temp_out, TAG_BTRFS_FILE);
                 return READ_ERROR;
             }
 
-            memcpy(out, temp_out + offset, size);
+            shift = physical & (512 - 1);
```
Comment (SergeGautherie) [inline boot/freeldr/freeldr/lib/fs/btrfs.c]: Need/Want a u64-to-u32 cast?

**Comment by SergeGautherie** [comment]
> capable of loading Server 2003

Hermès, to be explicit, we're talking about 'Windows on Btrfs' here, aren't we?

> differences wrt. loading XP

Mark, for what it's worth, would you know which (file) read (size) is failing?

--

Using less memory seems good anyway, assuming doing 2 `disk_read()` is an acceptable compromise...

**Comment by HBelusca** [comment]
> > capable of loading Server 2003
> 
> Hermès, to be explicit, we're talking about 'Windows on Btrfs' here, aren't we?
> 

Correct, but see this: https://youtu.be/-5E2CHmHEUs?t=31

**Comment by HBelusca** [comment]
> > differences wrt. loading XP
> 
> Mark, for what it's worth, would you know which (file) read (size) is failing?

Here it would be of help to enable all freeldr debugging channels (press F8 when in the boot menu -> freeldr debugging -> `+all`) (these are output to serial port), then looking at the results.

File: boot/freeldr/freeldr/lib/fs/btrfs.c (hunk 1 above)
Comment (Extravert-ir) [reply boot/freeldr/freeldr/lib/fs/btrfs.c]: The stack is small here

**Comment by maharmstone** [comment]
Debug log: [freeldr-winxp.txt](https://github.com/reactos/reactos/files/3721992/freeldr-winxp.txt). This is against fd291f1771ee609f87c7684ef22ebafcfd53d21e in https://github.com/maharmstone/reactos/tree/win7-freeldr, though I got the same results with the current upstream - the debug log was taking forever without the WinLdrScanRegistry patch.

This was while loading ntoskrnl.exe. The issue is that FrLdrHeapAllocateEx has a maximum size of MAXUSHORT * sizeof(HEAP_BLOCK) = 0x7fff8, and it was trying to allocate 0x8c8b8.

File: boot/freeldr/freeldr/lib/fs/btrfs.c (hunk 2 above)
Comment (tkreuzer) [inline boot/freeldr/freeldr/lib/fs/btrfs.c]: Isn't there some alignment macro for that?

File: boot/freeldr/freeldr/lib/fs/btrfs.c (hunk 3 above)
Comment (tkreuzer) [reply boot/freeldr/freeldr/lib/fs/btrfs.c]: Yes, please make sure that doesn't introduce a warning on MSVC builds.

**Comment by tkreuzer** [comment]
> 
> 
> We're still in real mode, where memory is at a premium, when we call BtrFsRead, and I was getting out of memory errors. This changes it so that we only allocate a temp buffer for the first 512-byte sector, and read normally from then on.

We're not in real mode, only the BIOS does a real mode access and thus requires a low address, which is what the DISKREADBUFFER is for. The temp heap is not guaranteed to be within the low 1MB. You were running out of memory, because the temp heap is simply not large enough to allocate buffers for entire files.

**Comment by maharmstone** [comment]
I've pushed a version taking into account the suggestions above.

**Pull request merged**